File: solutions/go/all-your-base/1/all_your_base.go
package allyourbase

import (
    "errors"
    "math"
)

func ConvertToBase(inputBase int, inputDigits []int, outputBase int) ([]int, error) {
	if inputBase < 2 {
        return nil, errors.New("input base must be >= 2")
    }
    if outputBase < 2 {
        return nil, errors.New("output base must be >= 2")
    }
    num, err := ToDecimal(inputBase, inputDigits)
    if err != nil {
        return nil, err
    }
    return FromDecimal(outputBase, num), nil
}

func ToDecimal(inputBase int, inputDigits []int) (int, error) {
    res := 0
    n := len(inputDigits)
    
    for i, x := range inputDigits {
        if x < 0 || x >= inputBase {
            return 0, errors.New("all digits must satisfy 0 <= d < input base")
        }
        res += x * int(math.Pow(float64(inputBase), float64(n - i - 1)))
    }
    return res, nil
}

func FromDecimal(outputBase int, inputDigits int) []int {
    res := []int{}
    if inputDigits == 0 {
        return []int{0}
    }
    for inputDigits > 0 {
        res = append(res, inputDigits % outputBase)
        inputDigits /= outputBase
    }
    n := len(res)
    for i := 0; i < n / 2; i++ {
        j := n - i - 1
        res[i], res[j] = res[j], res[i]
    }
    return res
}

File: solutions/python/all-your-base/1/all_your_base.py
def rebase(input_base, digits, output_base):
    if input_base < 2:
        raise ValueError("input base must be >= 2")
    if output_base < 2:
        raise ValueError("output base must be >= 2")
    num = toDecimal(input_base, digits)
    return fromDecimal(output_base, num)

def toDecimal(input_base, digits):
    res = 0
    n = len(digits)

    for i, x in enumerate(digits):
        if x < 0 or x >= input_base:
            raise ValueError("all digits must satisfy 0 <= d < input base")
        res += x * (input_base ** (n - i - 1))

    return res

def fromDecimal(output_base, num):
    res = []
    if num == 0:
        return [0]
    while num > 0:
        res.append(num % output_base)
        num //= output_base
    n = len(res)

    for i in range(n // 2):
        j = n - i - 1
        res[i], res[j] = res[j], res[i]

    return res